File: src/main/java/discretemath/bool/operator/Operators.java
package discretemath.bool.operator;

import discretemath.bool.expression.atomic.AtomicBooleanExpression;
import discretemath.bool.expression.BooleanExpression;
import discretemath.bool.expression.compound.CompoundBooleanExpression;
import discretemath.bool.expression.compound.ExpressionTreeNode;
import discretemath.bool.expression.compound.OperatorNode;
import discretemath.bool.expression.compound.ValueNode;
import discretemath.bool.expression.exception.InvalidArgumentCountException;

public enum Operators implements Operator {

	AND {
		@Override
		public boolean evaluate(boolean... values) {
			InvalidArgumentCountException.checkGTE(0, values.length);

			for (boolean value : values) {
				if (!value) {
					return false;
				}
			}

			return true;
		}

		@Override
		public BooleanExpression applyTo(BooleanExpression... operands) {
			InvalidArgumentCountException.checkGTE(2, operands.length);

			ExpressionTreeNode[] operandNodes = ExpressionTreeNode.fromBooleanExpression(operands);

			return new CompoundBooleanExpression(OperatorNode.forAND(operandNodes));
		}

		@Override
		public BooleanExpression applyUsing(FunctionallyCompleteOperatorsSet operatorsSet,
											BooleanExpression... operands) throws InvalidArgumentCountException, UnsupportedOperationException {
			InvalidArgumentCountException.checkExact(2, operands.length);

			BooleanExpression[] expressedExpressions = Operators.expressViaOperators(operatorsSet, operands);
			ExpressionTreeNode[] operandNodes = ExpressionTreeNode.fromBooleanExpression(expressedExpressions);

			ExpressionTreeNode firstExpressionRootNode = operandNodes[0];
			ExpressionTreeNode secondExpressionRootNode = operandNodes[1];

			return switch (operatorsSet) {
				case OR_NOT -> new CompoundBooleanExpression(OperatorNode.forNegation(
						OperatorNode.forOR(OperatorNode.forNegation(firstExpressionRootNode),
								OperatorNode.forNegation(secondExpressionRootNode))));
				case SINGLE_NAND -> new CompoundBooleanExpression(
						OperatorNode.forNAND(OperatorNode.forNAND(firstExpressionRootNode, secondExpressionRootNode),
								OperatorNode.forNAND(firstExpressionRootNode, secondExpressionRootNode)));
				case SINGLE_NOR -> new CompoundBooleanExpression(
						OperatorNode.forNOR(OperatorNode.forNOR(firstExpressionRootNode, firstExpressionRootNode),
								OperatorNode.forNOR(secondExpressionRootNode, secondExpressionRootNode)));
				default -> throw new UnsupportedOperationException();
			};
		}

		@Override
		public String toString(BooleanExpression... operands) throws InvalidArgumentCountException {
			InvalidArgumentCountException.checkGTE(2, operands.length);

			return OperatorsStringUtils.buildNAryOperatorString("", operands);
		}
	}, OR {
		@Override
		public boolean evaluate(boolean... values) {
			InvalidArgumentCountException.checkGTE(0, values.length);

			for (boolean value : values) {
				if (value) {
					return true;
				}
			}

			return false;
		}

		@Override
		public BooleanExpression applyTo(BooleanExpression... operands) {
			InvalidArgumentCountException.checkGTE(2, operands.length);

			ExpressionTreeNode[] operandNodes = ExpressionTreeNode.fromBooleanExpression(operands);

			return new CompoundBooleanExpression(OperatorNode.forOR(operandNodes));
		}

		@Override
		public BooleanExpression applyUsing(FunctionallyCompleteOperatorsSet operatorsSet,
											BooleanExpression... operands) throws InvalidArgumentCountException, UnsupportedOperationException {
			InvalidArgumentCountException.checkExact(2, operands.length);

			BooleanExpression[] expressedExpressions = Operators.expressViaOperators(operatorsSet, operands);
			ExpressionTreeNode[] operandNodes = ExpressionTreeNode.fromBooleanExpression(expressedExpressions);

			ExpressionTreeNode firstExpressionRootNode = operandNodes[0];
			ExpressionTreeNode secondExpressionRootNode = operandNodes[1];

			return switch (operatorsSet) {
				case AND_NOT -> new CompoundBooleanExpression(OperatorNode.forNegation(
						OperatorNode.forAND(OperatorNode.forNegation(firstExpressionRootNode),
								OperatorNode.forNegation(secondExpressionRootNode))));
				case SINGLE_NAND -> new CompoundBooleanExpression(
						OperatorNode.forNAND(OperatorNode.forNAND(firstExpressionRootNode, firstExpressionRootNode),
								OperatorNode.forNAND(secondExpressionRootNode, secondExpressionRootNode)));
				case SINGLE_NOR -> new CompoundBooleanExpression(
						OperatorNode.forNOR(OperatorNode.forNOR(firstExpressionRootNode, secondExpressionRootNode),
								OperatorNode.forNOR(firstExpressionRootNode, secondExpressionRootNode)));
				default -> throw new UnsupportedOperationException();
			};
		}

		@Override
		public String toString(BooleanExpression... operands) throws InvalidArgumentCountException {
			InvalidArgumentCountException.checkGTE(2, operands.length);

			return OperatorsStringUtils.buildNAryOperatorString(" + ", operands);
		}

	}, NOT {
		@Override
		public boolean evaluate(boolean... values) {
			InvalidArgumentCountException.checkExact(1, values.length);

			return !values[0];
		}

		@Override
		public BooleanExpression applyTo(BooleanExpression... operands) {
			InvalidArgumentCountException.checkExact(1, operands.length);

			BooleanExpression operand = operands[0];

			ExpressionTreeNode treeNode;
			if (operand instanceof AtomicBooleanExpression expression) {
				treeNode = new ValueNode(expression);
			} else if (operand instanceof CompoundBooleanExpression expression) {
				treeNode = expression.getRootNode();
			} else {
				throw new UnsupportedOperationException();
			}

			return new CompoundBooleanExpression(OperatorNode.forNegation(treeNode));
		}

		@Override
		public BooleanExpression applyUsing(FunctionallyCompleteOperatorsSet operatorsSet,
											BooleanExpression... operands) throws InvalidArgumentCountException, UnsupportedOperationException {
			InvalidArgumentCountException.checkExact(1, operands.length);

			BooleanExpression booleanExpression = operands[0].expressViaOperators(operatorsSet);
			ExpressionTreeNode treeNode = ExpressionTreeNode.fromBooleanExpression(booleanExpression);

			if (operatorsSet.contains(this)) {
				return applyTo(booleanExpression);
			} else {
				return switch (operatorsSet) {
					case SINGLE_NAND -> new CompoundBooleanExpression(OperatorNode.forNAND(treeNode, treeNode));
					case SINGLE_NOR -> new CompoundBooleanExpression(OperatorNode.forNOR(treeNode, treeNode));
					default -> throw new UnsupportedOperationException();
				};
			}
		}

		@Override
		public String toString(BooleanExpression... operands) throws InvalidArgumentCountException {
			InvalidArgumentCountException.checkExact(1, operands.length);

			return OperatorsStringUtils.wrapWithParentheses(operands[0]) + "′";
		}
	}, NAND {
		@Override
		public boolean evaluate(boolean... values) {
			InvalidArgumentCountException.checkGTE(0, values.length);

			for (boolean value : values) {
				if (!value) {
					return true;
				}
			}

			return false;
		}

		@Override
		public BooleanExpression applyTo(BooleanExpression... operands) {
			InvalidArgumentCountException.checkExact(2, operands.length);

			ExpressionTreeNode firstExpressionRootNode = ExpressionTreeNode.fromBooleanExpression(operands[0]);
			ExpressionTreeNode secondExpressionRootNode = ExpressionTreeNode.fromBooleanExpression(operands[1]);

			return new CompoundBooleanExpression(
					OperatorNode.forNAND(firstExpressionRootNode, secondExpressionRootNode));
		}

		@Override
		public BooleanExpression applyUsing(FunctionallyCompleteOperatorsSet operatorsSet,
											BooleanExpression... operands) throws InvalidArgumentCountException, UnsupportedOperationException {
			throw new UnsupportedOperationException();
		}

		@Override
		public String toString(BooleanExpression... operands) throws InvalidArgumentCountException {
			InvalidArgumentCountException.checkExact(2, operands.length);

			return OperatorsStringUtils.buildNAryOperatorString(" | ", operands);
		}
	}, NOR {
		@Override
		public boolean evaluate(boolean... values) {
			InvalidArgumentCountException.checkGTE(0, values.length);

			for (boolean value : values) {
				if (value) {
					return false;
				}
			}

			return true;
		}

		@Override
		public BooleanExpression applyTo(BooleanExpression... operands) {
			InvalidArgumentCountException.checkExact(2, operands.length);

			ExpressionTreeNode firstExpressionRootNode = ExpressionTreeNode.fromBooleanExpression(operands[0]);
			ExpressionTreeNode secondExpressionRootNode = ExpressionTreeNode.fromBooleanExpression(operands[1]);

			return new CompoundBooleanExpression(
					OperatorNode.forNOR(firstExpressionRootNode, secondExpressionRootNode));
		}

		@Override
		public BooleanExpression applyUsing(FunctionallyCompleteOperatorsSet operatorsSet,
											BooleanExpression... operands) throws InvalidArgumentCountException, UnsupportedOperationException {
			throw new UnsupportedOperationException();
		}

		@Override
		public String toString(BooleanExpression... operands) throws InvalidArgumentCountException {
			InvalidArgumentCountException.checkExact(2, operands.length);

			return OperatorsStringUtils.buildNAryOperatorString(" ↓ ", operands);
		}
	};

	private static BooleanExpression[] expressViaOperators(FunctionallyCompleteOperatorsSet operatorsSet,
														   BooleanExpression... expressions) {
		BooleanExpression[] newExpressions = new BooleanExpression[expressions.length];
		for (int i = 0; i < expressions.length; i++) {
			BooleanExpression expression = expressions[i];

			newExpressions[i] = expression.expressViaOperators(operatorsSet);
		}

		return newExpressions;
	}
}

File: src/main/scala/discretemath/bool/operator/Operators.scala
package discretemath.bool.operator

import discretemath.bool.expression.BooleanExpression
import discretemath.bool.expression.compound.{CompoundBooleanExpression, ExpressionTreeNode, OperatorNode}
import discretemath.bool.operator.FunctionallyCompleteOperatorsSet._

object Operators {

  object AND extends Operator {

    override def evaluate(values: Seq[Boolean]): Boolean = {
      assert(values.nonEmpty)

      for (value <- values) {
        if (!value) return false
      }
      true
    }

    override def inject(operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      injectOperator(this, operands)
    }

    @throws[UnsupportedOperationException]
    override def injectUsing(operatorsSet: FunctionallyCompleteOperatorsSet, operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      val expressedExpressions = expressViaOperators(operatorsSet, operands)

      if (operatorsSet.contains(this)) {
        return inject(expressedExpressions)
      }

      val operandNodes = ExpressionTreeNode.fromBooleanExpression(expressedExpressions)

      val rootNode = operatorsSet match {
        case OR_NOT => OperatorNode.not(OperatorNode(OR, operandNodes.map(node => OperatorNode.not(node))))
        case ONLY_NOR => OperatorNode(NOR, operandNodes.map(node => OperatorNode * (NOR, node, node)))
        case ONLY_NAND =>
          val nandCombined = OperatorNode(NAND, operandNodes)
          OperatorNode * (NAND, nandCombined, nandCombined)
        case _ => throw new UnsupportedOperationException
      }

      CompoundBooleanExpression(rootNode)
    }

    override def toString(operands: Seq[BooleanExpression]): String = {
      assert(operands.length >= 2)

      buildNAryOperatorString("", operands)
    }
  }

  object OR extends Operator {
    override def evaluate(values: Seq[Boolean]): Boolean = {
      assert(values.nonEmpty)

      for (value <- values) {
        if (value) return true
      }
      false
    }

    override def inject(operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      injectOperator(this, operands)
    }

    @throws[UnsupportedOperationException]
    override def injectUsing(operatorsSet: FunctionallyCompleteOperatorsSet, operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      val expressedExpressions = expressViaOperators(operatorsSet, operands)

      if (operatorsSet.contains(this)) {
        return inject(expressedExpressions)
      }

      val operandNodes = ExpressionTreeNode.fromBooleanExpression(expressedExpressions)

      val rootNode = operatorsSet match {
        case AND_NOT => OperatorNode.not(OperatorNode(AND, operandNodes.map(node => OperatorNode.not(node))))
        case ONLY_NAND => OperatorNode(NAND, operandNodes.map(node => OperatorNode * (NAND, node, node)))
        case ONLY_NOR =>
          val nandCombined = OperatorNode(NOR, operandNodes)
          OperatorNode * (NOR, nandCombined, nandCombined)
        case _ => throw new UnsupportedOperationException
      }

      CompoundBooleanExpression(rootNode)
    }

    override def toString(operands: Seq[BooleanExpression]): String = {
      assert(operands.length >= 2)

      buildNAryOperatorString(" + ", operands)
    }
  }

  object NOT extends Operator {

    override def evaluate(values: Seq[Boolean]): Boolean = {
      assert(values.length == 1)

      return !values.head
    }

    override def inject(operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length == 1)

      injectOperator(this, operands)
    }

    @throws[UnsupportedOperationException]
    override def injectUsing(operatorsSet: FunctionallyCompleteOperatorsSet, operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length == 1)

      val booleanExpression = operands.head.expressViaOperators(operatorsSet)

      if (operatorsSet.contains(this)) {
        return inject(Seq(booleanExpression))
      }

      val node = ExpressionTreeNode.fromBooleanExpression(booleanExpression)

      return operatorsSet match {
        case ONLY_NAND => new CompoundBooleanExpression(OperatorNode * (NAND, node, node))
        case ONLY_NOR => new CompoundBooleanExpression(OperatorNode * (NOR, node, node))
        case _ => throw new UnsupportedOperationException();
      }
    }

    override def toString(operands: Seq[BooleanExpression]): String = {
      assert(operands.length == 1)

      return "~" + wrapWithParentheses(operands.head)
    }
  }

  object NAND extends Operator {
    override def evaluate(values: Seq[Boolean]): Boolean = {
      assert(values.nonEmpty)

      for (value <- values) {
        if (!value) return true
      }
      return false
    }

    override def inject(operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      injectOperator(this, operands)
    }

    @throws[UnsupportedOperationException]
    override def injectUsing(operatorsSet: FunctionallyCompleteOperatorsSet, operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      val expressedExpressions = expressViaOperators(operatorsSet, operands)

      if (operatorsSet.contains(this)) {
        return inject(expressedExpressions)
      }

      throw new NotImplementedError()
    }

    override def toString(operands: Seq[BooleanExpression]): String = {
      assert(operands.length >= 2)

      return buildNAryOperatorString(" ↑ ", operands)
    }
  }

  object NOR extends Operator {
    override def evaluate(values: Seq[Boolean]): Boolean = {
      assert(values.nonEmpty)

      for (value <- values) {
        if (value) return false
      }
      return true
    }

    override def inject(operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      injectOperator(this, operands)
    }

    @throws[UnsupportedOperationException]
    override def injectUsing(operatorsSet: FunctionallyCompleteOperatorsSet, operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      val expressedExpressions = expressViaOperators(operatorsSet, operands)

      if (operatorsSet.contains(this)) {
        return inject(expressedExpressions)
      }

      throw new NotImplementedError()
    }

    override def toString(operands: Seq[BooleanExpression]): String = {
      assert(operands.length >= 2)

      return buildNAryOperatorString(" ↓ ", operands)
    }
  }

  object XOR extends Operator {
    override def evaluate(values: Seq[Boolean]): Boolean = {
      assert(values.nonEmpty)

      var result = values.head

      for (i <- 1 until values.length) {
        val value = values(i)
        result = result ^ value
      }

      return result
    }

    override def inject(operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      injectOperator(this, operands)
    }

    @throws[UnsupportedOperationException]
    override def injectUsing(operatorsSet: FunctionallyCompleteOperatorsSet, operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length >= 2)

      val expressedExpressions = expressViaOperators(operatorsSet, operands)

      if (operatorsSet.contains(this)) {
        return inject(expressedExpressions)
      }

      throw new NotImplementedError()
    }

    override def toString(operands: Seq[BooleanExpression]): String = {
      assert(operands.length >= 2)

      return buildNAryOperatorString(" ⊕ ", operands);
    }
  }

  object CONDITIONAL extends Operator {
    override def evaluate(values: Seq[Boolean]): Boolean = {
      return values match {
        case Seq(hypothesis, conclusion) => !hypothesis || conclusion
        case _ => throw new AssertionError()
      }
    }

    override def inject(operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length == 2)

      injectOperator(this, operands)
    }

    @throws[UnsupportedOperationException]
    override def injectUsing(operatorsSet: FunctionallyCompleteOperatorsSet, operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length == 2)

      val expressedExpressions = expressViaOperators(operatorsSet, operands)

      if (operatorsSet.contains(this)) {
        return inject(expressedExpressions)
      }

      throw new NotImplementedError()
    }

    override def toString(operands: Seq[BooleanExpression]): String = {
      assert(operands.length == 2)

      return buildNAryOperatorString(" → ", operands)
    }
  }

  object BICONDITIONAL extends Operator {
    override def evaluate(values: Seq[Boolean]): Boolean = {
      return values match {
        case Seq(conclusion1, conclusion2) => conclusion1 == conclusion2
        case _ => throw new AssertionError()
      }
    }

    override def inject(operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length == 2)

      injectOperator(this, operands)
    }

    @throws[UnsupportedOperationException]
    override def injectUsing(operatorsSet: FunctionallyCompleteOperatorsSet, operands: Seq[BooleanExpression]): BooleanExpression = {
      assert(operands.length == 2)

      val expressedExpressions = expressViaOperators(operatorsSet, operands)

      if (operatorsSet.contains(this)) {
        return inject(expressedExpressions)
      }

      throw new NotImplementedError()
    }

    override def toString(operands: Seq[BooleanExpression]): String = {
      assert(operands.length == 2)

      return buildNAryOperatorString(" ⟷ ", operands)
    }
  }
}
